lib/netext/dialer.go

```diff
@@ -44,13 +49,162 @@ type Dialer struct {
 
 	BytesRead    int64
 	BytesWritten int64
+
+	dnsCache DNSCache
+	ttl      time.Duration
+	selecter selecter
 }
 
 // NewDialer constructs a new Dialer with the given DNS resolver.
-func NewDialer(dialer net.Dialer, resolver Resolver) *Dialer {
-	return &Dialer{
-		Dialer:   dialer,
-		Resolver: resolver,
+func NewDialer(
+	dialer net.Dialer, blacklist []*lib.IPNet, hosts map[string]*lib.HostAddress,
+	dnsConfig lib.DNSConfig,
+	// TODO take DNSCache so it's shared between VUs
+) *Dialer {
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
+
+	ttl, err := parseTTL(dnsConfig.TTL.String)
+	if err != nil {
+		panic(err) // TODO fix
+	}
+	strategy := dnsConfig.Strategy.DNSStrategy
+	if !strategy.IsADNSStrategy() {
+		strategy = lib.DefaultDNSConfig().Strategy.DNSStrategy
+	}
+	d := &Dialer{
+		Blacklist: blacklist,
+		Hosts:     hosts,
+		selecter: selecter{
+			strategy:   strategy,
+			rrm:        &sync.Mutex{},
+			rand:       r,
+			roundRobin: make(map[string]uint8),
+		},
+		ttl: ttl,
+		dnsCache: DNSCache{
+			v4: &dnsCache{
+				RWMutex: sync.RWMutex{},
+				cache:   make(map[string]*cacheRecord),
+			},
+			v6: &dnsCache{
+				RWMutex: sync.RWMutex{},
+				cache:   make(map[string]*cacheRecord),
+			},
```

Comment:
Why are separate caches needed for v4 and v6?

Reply:
We get separate requests for them so it was easier to have them separate